// @loopback/rest/dist/providers/send.provider.js
"use strict";
// Copyright IBM Corp. 2018,2020. All Rights Reserved.
// Node module: @loopback/rest
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendResponseMiddlewareProvider = exports.SendProvider = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@loopback/core");
const express_1 = require("@loopback/express");
const keys_1 = require("../keys");
const sequence_1 = require("../sequence");
const writer_1 = require("../writer");
/**
 * Provides the function that populates the response object with
 * the results of the operation.
 *
 * @returns The handler function that will populate the
 * response with operation results.
 */
class SendProvider {
    value() {
        return writer_1.writeResultToResponse;
    }
}
exports.SendProvider = SendProvider;
let SendResponseMiddlewareProvider = class SendResponseMiddlewareProvider {
    constructor(send, reject) {
        this.send = send;
        this.reject = reject;
    }
    value() {
        return async (ctx, next) => {
            try {
                /**
                 * Invoke downstream middleware to produce the result
                 */
                const result = await next();
                /**
                 * Write the result to HTTP response
                 */
                this.send(ctx.response, result);
            }
            catch (err) {
                /**
                 * Write the error to HTTP response
                 */
                this.reject(ctx, err);
            }
        };
    }
};
SendResponseMiddlewareProvider = tslib_1.__decorate([
    core_1.bind(express_1.asMiddleware({
        group: sequence_1.RestMiddlewareGroups.SEND_RESPONSE,
        downstreamGroups: [
            sequence_1.RestMiddlewareGroups.CORS,
            sequence_1.RestMiddlewareGroups.INVOKE_METHOD,
        ],
        chain: keys_1.RestTags.REST_MIDDLEWARE_CHAIN,
    })),
    tslib_1.__param(0, core_1.inject(keys_1.RestBindings.SequenceActions.SEND)),
    tslib_1.__param(1, core_1.inject(keys_1.RestBindings.SequenceActions.REJECT)),
    tslib_1.__metadata("design:paramtypes", [Function, Function])
], SendResponseMiddlewareProvider);
exports.SendResponseMiddlewareProvider = SendResponseMiddlewareProvider;
